contracts/src/utils/airdrop/merkle/MerkleAirdrop.sol

```diff
@@ -0,0 +1,120 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+// interfaces
+import {IMerkleAirdrop} from "./IMerkleAirdrop.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+// libraries
+import {MerkleProof} from "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
+
+// contracts
+import {EIP712Base} from "contracts/src/diamond/utils/cryptography/signature/EIP712Base.sol";
+import {ECDSA} from "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {Facet} from "contracts/src/diamond/facets/Facet.sol";
+
+contract MerkleAirdrop is IMerkleAirdrop, EIP712Base, Facet {
+  // keccak256("AirdropClaim(address account,uint256 amount)");
+  bytes32 private constant MESSAGE_TYPEHASH =
+    0xaa726e564e52b64144617a6a46c42e8b763d4d224ca1a3e13c1491f8a4763a23;
+
+  function __MerkleAirdrop_init(
+    bytes32 merkleRoot,
+    IERC20 token
+  ) external initializer {
+    _addInterface(type(IMerkleAirdrop).interfaceId);
+    __MerkleAirdrop_init_unchained(merkleRoot, token);
+  }
+
+  /// @inheritdoc IMerkleAirdrop
+  function getMerkleRoot() public view returns (bytes32) {
+    return MerkleAirdropStorage.layout().merkleRoot;
+  }
+
+  /// @inheritdoc IMerkleAirdrop
+  function getToken() public view returns (IERC20) {
+    return MerkleAirdropStorage.layout().token;
+  }
+
+  /// @inheritdoc IMerkleAirdrop
+  function getMessageHash(
+    address account,
+    uint256 amount
+  ) public view returns (bytes32) {
+    return
+      _hashTypedDataV4(
+        keccak256(abi.encode(MESSAGE_TYPEHASH, AirdropClaim(account, amount)))
+      );
+  }
+
+  /// @inheritdoc IMerkleAirdrop
+  function claim(
+    address account,
+    uint256 amount,
+    bytes32[] calldata merkleProof,
+    bytes memory signature
+  ) external {
+    MerkleAirdropStorage.Layout storage ds = MerkleAirdropStorage.layout();
+
+    if (ds.claimed[account]) {
+      revert MerkleAirdrop__AlreadyClaimed();
+    }
+
+    _validateSignature(account, getMessageHash(account, amount), signature);
+
+    // verify merkle proof
+    //should we use bytes32 leaf = keccak256(bytes.concat(keccak256(abi.encode(account, amount))));
+    bytes32 leaf = keccak256(abi.encodePacked(account, amount));
+
+    if (!MerkleProof.verify(merkleProof, ds.merkleRoot, leaf)) {
+      revert MerkleAirdrop__InvalidProof();
+    }
+
+    ds.claimed[account] = true;
+    emit Claimed(account, amount);
+    SafeERC20.safeTransfer(ds.token, account, amount);
```

Comment:
```suggestion
    SafeTransferLib.safeTransfer(address(ds.token), account, amount);
```
We can potentially allow claiming to a different address provided a signature from `account`.